core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -1153,39 +1153,29 @@ private[sql] final class PreprocessTableInsertOrPut(conf: SQLConf)
     // ResolveRelations, no such special rule has been added for PUT
     case d@DeleteFromTable(table, child) if table.resolved && child.resolved =>
       EliminateSubqueryAliases(table) match {
-        case l@LogicalRelation(dr: DeletableRelation, _, _) =>
+        case l@LogicalRelation(dr: MutableRelation, _, _) =>
           def comp(a: Attribute, targetCol: String): Boolean = a match {
             case ref: AttributeReference => targetCol.equals(ref.name.toUpperCase)
           }
 
-          val expectedOutput = l.output
-          if (!child.output.forall(a => expectedOutput.exists(e => comp(a, e.name.toUpperCase)))) {
-            throw new AnalysisException(s"$l requires that the query in the " +
-                "WHERE clause of the DELETE FROM statement " +
-                "generates the same column name(s) as in its schema but found " +
-                s"${child.output.mkString(",")} instead.")
-          }
-          l match {
-            case LogicalRelation(ps: PartitionedDataSourceScan, _, _) =>
-              if (!ps.partitionColumns.forall(a => child.output.exists(e =>
-                comp(e, a.toUpperCase)))) {
-                throw new AnalysisException(s"${child.output.mkString(",")}" +
-                    s" columns in the WHERE clause of the DELETE FROM statement must " +
-                    s"have all the parititioning column(s) ${ps.partitionColumns.mkString(",")}.")
-              }
-            case _ =>
+          val childAttributes = child.output.map(ar => (ar.name.toUpperCase(), ar)).toMap
+          val keyColumns = dr.getPrimaryKeyColumns
+
+          val childOutput = try {
+            keyColumns.map(_.toUpperCase)
+                .map(childAttributes(_))
+          } catch {
+            case _: NoSuchElementException =>
+              throw new AnalysisException(s"$l requires that the query in the " +
+                  "WHERE clause of the DELETE FROM statement " +
+                  s"must have all the key column(s) ${keyColumns.mkString(",")} but found " +
+                  s"${child.output.mkString(",")} instead.")
           }
-          castAndRenameChildOutputForPut(d, expectedOutput, dr, l, child)
 
-        case l@LogicalRelation(dr: MutableRelation, _, _) =>
-          val expectedOutput = l.output
-          if (child.output.length != expectedOutput.length) {
-            throw new AnalysisException(s"$l requires that the query in the " +
-                "WHERE clause of the DELETE FROM statement " +
-                "generates the same number of column(s) as in its schema but found " +
-                s"${child.output.mkString(",")} instead.")
-          }
-          castAndRenameChildOutputForPut(d, expectedOutput, dr, l, child)
+          val relationAttributes = l.output.map(ar => (ar.name.toUpperCase(), ar)).toMap
+          val expectedOutput = keyColumns.map(_.toUpperCase).map(relationAttributes(_))
```

Comment:
The standard way to resolve column names among Attributes both here and above is using LogicalPlan.resolveQuoted like:
```scala
val expectedOutput = keyColumns.map(col => l.resolveQuoted(col, analysis.caseInsensitiveResolution) match {
  case Some(a: Attribute) => a
  case _ => throw ...
})
```

(here analysis is package object in Spark's org.apache.spark.sql.catalyst)
Similarly for child.resolveQuoted() above.

Reply:
To clarify: resolveQuoted matches against the entire tree so returns Option[NamedExpression] which will be an alias in case the matching column is somewhere below or an expression alias itself. However, we need it to be always a direct attribute and hence the match against Some(a: Attribute).